provider/oracle/instance.go

```diff
@@ -0,0 +1,411 @@
+	name string
+	// status holds the status of the instance
+	status instance.InstanceStatus
+	// machine will hold the raw instance details obtained from
+	// the provider
+	machine response.Instance
+	// arch will hold the architecture information of the instance
+	arch      *string
+	instType  *instances.InstanceType
+	mutex     *sync.Mutex
+	env       *oracleEnviron
+	machineId string
+}
+
+// hardwareCharacteristics returns the hardware characteristics of the current
+// instance
+func (o *oracleInstance) hardwareCharacteristics() *instance.HardwareCharacteristics {
+	if o.arch == nil {
+		return nil
+	}
+
+	hc := &instance.HardwareCharacteristics{Arch: o.arch}
+	if o.instType != nil {
+		hc.Mem = &o.instType.Mem
+		hc.RootDisk = &o.instType.RootDisk
+		hc.CpuCores = &o.instType.CpuCores
+	}
+
+	return hc
+}
+
+// newInstance returns a new oracleInstance
+func newInstance(params response.Instance, env *oracleEnviron) (*oracleInstance, error) {
+	if params.Name == "" {
+		return nil, errors.New(
+			"Instance response does not contain a name",
+		)
+	}
+	//gsamfira: there must be a better way to do this.
+	splitMachineName := strings.Split(params.Label, "-")
+	machineId := splitMachineName[len(splitMachineName)-1]
+	mutex := &sync.Mutex{}
+	instance := &oracleInstance{
+		name: params.Name,
+		status: instance.InstanceStatus{
+			Status:  status.Status(params.State),
+			Message: "",
+		},
+		machine:   params,
+		mutex:     mutex,
+		env:       env,
+		machineId: machineId,
+	}
+
+	return instance, nil
+}
+
+// Id is defined on the instance.Instance interface.
+func (o *oracleInstance) Id() instance.Id {
+	if o.machine.Name != "" {
+		return instance.Id(o.machine.Name)
+	}
+
+	return instance.Id(o.name)
+}
+
+// Status is defined on the instance.Instance interface.
+func (o *oracleInstance) Status() instance.InstanceStatus {
+	return o.status
+}
+
+// StorageAttachments returns the storage that was attached in the moment
+// of instance creation. This storage cannot be detached dynamically.
+// this is also needed if you wish to determine the free disk index
+// you can use when attaching a new disk
+func (o *oracleInstance) StorageAttachments() []ociResponse.Storage {
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+	return o.machine.Storage_attachments
+}
+
+// refresh refreshes the instance raw details from the oracle api
+// this method is mutex protected
+func (o *oracleInstance) refresh() error {
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	machine, err := o.env.client.InstanceDetails(o.name)
+	// if the request failed for any reason
+	// we should not update the information and
+	// let the old one persist
+	if err != nil {
+		return err
+	}
+
+	o.machine = machine
+	return nil
+}
+
+// waitForMachineStatus will ping the machine status until the timeout
+// duration is reached or an error appeared
+func (o *oracleInstance) waitForMachineStatus(state ociCommon.InstanceState, timeout time.Duration) error {
+	errChan := make(chan error)
```

Comment:
this can be simplified to a loop & select at the top level, without the additional goroutine:

```go
ticker := time.NewTicker(10 * time.Second)
defer ticker.Stop()
timer := time.NewTimer(timeout)
defer timer.Stop()

for {
    select {
    case <-ticker.C:
        // refresh, check machine status
    case <-timer.C:
        // return timeout error
    }
}
```